Add typing and annotations to the point of mypy's satisfaction

ndevenish opened the pull request:
This adds typing stubs and resolves mypy typing complaints to the point where `mypy  --ignore-missing-imports` shows no errors. This is filtering out some large portion of important dependencies like `h5py`, `matplotlib` don't have typing information that mypy can use.

Tasks done in this PR:
- Added typing stubs for dxtbx C++ extension modules. Using these involves making dxtbx attempt to import these modules from the dxtbx root first - rather than populating the stubs at the system-global level
- Ignore lines where typing gets confused e.g. module backup imports
- Add typing annotations where mypy cannot infer the typing enough to pass
- Add asserts where the code indicates an assumption is made, but otherwise relies on the input types to be properly formed e.g. imageset metadata where the filename is `Optional[str]`, but is assumed/guaranteed to have a value at the point of processing

Obviously, `--ignore-missing-imports` covers a very large proportion of imported cctbx_project modules and feature. To resolve this I used a mix of autogeneration and manual annotation, wrote an (very informal) stub library over the parts of cctbx that we use here: https://github.com/ndevenish/types-cctbx_project. This reduces the run without `--ignore-missing-imports` from 344 to 82 import errors. This stub library should mostly be okay for the existence of methods and function arguments, but does mostly type individual arguments and returns as `Any`. **This stub module is completely optional and all typing with ignored imports passes exactly the same without it**.

ndevenish commented:
Turning draft to reflect the few odds and ends to polish off

Anthchirp commented on src/dxtbx/model/experiment_list.py:
```diff
@@ -1306,8 +1320,9 @@ def _create_imageset(records, format_class, format_kwargs=None):
-def _create_imagesequence(record, format_class, format_kwargs=None):
-    # type: (ImageMetadataRecord, Type[Format], Dict) -> dxtbx.imageset.ImageSequence
+def _create_imagesequence(
+    record: ImageMetadataRecord, format_class: Type[Format], format_kwargs: Dict = None
```
I'm surprised this isn't flagged as invalid annotation?

ndevenish replied:
Good catch. From https://mypy.readthedocs.io/en/stable/kinds_of_types.html#optional-types-and-the-none-type:
> You can use the --no-implicit-optional command-line option to stop treating arguments with a None default value as having an implicit Optional[...] type. It’s possible that this will become the default behavior in the future.

So we should turn this on.

Anthchirp commented on src/dxtbx/nexus/nxmx.py:
```diff
@@ -76,9 +87,12 @@ def find_classes(
         A list of matching nodes for each of the specified NX_class types.
     """
-    results = {nx_class: [] for nx_class in nx_classes}
+    results: Dict[Optional[str], List[h5py.Group]] = {
```
Picking one at random:
if we added the future import everywhere this could become
```suggestion
    results: dict[Optional[str], list[h5py.Group]] = {
```
following what we did in https://github.com/dials/data/pull/318

I think that'd be good practice as this is going to be the future anyway, so people don't see (copy around) outdated coding conventions.

ndevenish replied:
Interesting! I knew that it came in with Python 3.9, but obviously hadn't joined the dots that as long as you don't try to evaluate the annotation in-interpreter it of course works fine. Yes, I think is better to go to early.

Anthchirp reviewed (approved):
All in all great effort. You probably checked whether one could just declare the C++ types in the files directly rather than having the somewhat unclean `try import` block?

ndevenish commented:
Right, after discussion, I have:

- Added `from __future__ import annotations` to all files
- Using a custom type-only pyupgrade I have:
  - Replaced references to `Dict`, `List` etc with their generic bases (e.g. `list[int]`)
  - Replaced annotations using `Union[x, y]` with `x | y`
  - Replaced annotations `Optional[x]` with `x | None`

All of these changes are valid in 3.7, as long as you don't attempt to parse the type hints (with `typing.get_type_hints()`). Type checkers recognise the newer syntax even on older codebases.

ndevenish commented:
The comment regarding C++ types and duplicated import lines - yes, they could mostly just go straight into the importing module (for cases where extensions are not imported in multiple places to avoid circular references), however, I believe that this import structure is probably necessary if we want to e.g. distribute dxtbx without polluting the top-level python namespace - I think we've discussed this in the past. I can't immediately think of a simple, cleaner way to do this.

phyy-nx commented:
Hi, FYI I'm manually running the xfel_regression tests here using the current typing branch as of a few minutes ago:
https://dev.azure.com/cctbx/cctbx_project/_build/results?buildId=18942&view=results

This will take a few hours.  If there are more changes I can run these tests again.

phyy-nx commented:
Xfel regression tests passed :)
